Use of pydantic to resolve tags in static tests

marioevz opened the pull request:
## 🗒️ Description
Uses pydantic machinery to simplify parsing and resolution of tags in the static tests.

### New tags
`<create:my_contract:0x01>`: References contract `my_contract` and resolves to the calculated contract creation address of nonce `0x01` with `CREATE` or a contract creating tx.
`<create2:my_contract:0x01:0x012345abcd...>`: References contract `my_contract` and resolves to the calculated contract creation address of salt `0x01` and initcode `0x012345abcd...` using `CREATE2`.

See `CREATE2_HighNonceDelegatecallFiller.yml` for an example.

fselmo commented on src/ethereum_test_specs/static_state/common/tags.py:
```diff
@@ -0,0 +1,133 @@
+    @model_validator(mode="before")
+    @classmethod
+    def validate_from_string(cls, data: Any) -> Any:
+        """Validate the sender tag from string: <eoa:name:0x...>."""
```
This isn't always a sender tag, this should maybe just remove the word "sender"?

marioevz replied:
That's true, `eoa` should be better I think?

Edit: Now I see the position of this code snippet, I'll change it to something much more generic.

fselmo commented on src/ethereum_test_specs/static_state/common/tags.py:
```diff
@@ -0,0 +1,133 @@
+
+    def resolve(self, tags: TagDict) -> EOA:
+        """Resolve the tag."""
+        assert self.name in tags, f"Tag {self.name} not found in tags"
```
The sender isn't always included though the key may be. Will this be an issue? There are some tests where there is a key but no sender in `pre` and the transaction is expected to fail in that case.

marioevz replied:
I think as long as both the key tag and the sender tag have the same tag-name it should be ok, since the key is provided by `pre.fund_eoa` and not the other way around.

Edit: My comment is completely wrong, this is the actual solution:
https://github.com/marioevz/execution-spec-tests/blob/8efef46b09e0ca65448c9491f4a4d9b71f7dcd73/src/ethereum_test_specs/static_state/account.py#L132-L139

fselmo reviewed (commented):
I'm a big fan of the pydantic refactor but I don't see any changes in the script to convert hard-coded addresses to tags. I did a quick change for the `sender:key` to `eoa:sender` but everything is still failing for me. How did you test this? Can you give me a reproducible way to test what you changed here?

The script is at `scripts/convert_addresses.py`.

fselmo commented:
> I'm a big fan of the pydantic refactor but I don't see any changes in the script to convert hard-coded addresses to tags. I did a quick change for the `sender:key` to `eoa:sender` but everything is still failing for me. How did you test this? Can you give me a reproducible way to test what you changed here?


I did try the `CREATE2_HighNonceDelegatecallFiller.yml` and I do see that is working. But something like `uv run fill --fill-static-tests --clean tests/static -k stRandom` for example fails all tests. Does your approach have to go test by test and tag? 

If I try to tweak `tests/static/state_tests/stRandom/randomStatetest4Filler.json`, for example, it looks like things resolve but fails the test with unexpected nonce. Something doesn't seem correct there.

fselmo commented:
> If I try to tweak `tests/static/state_tests/stRandom/randomStatetest4Filler.json`, for example, it looks like things resolve but fails the test with unexpected nonce. Something doesn't seem correct there.

For this test, for example, I'm seeing the sender is actually still the hard-coded sender and I'm seeing a `nonce` of `1` in the `pre` for the coinbase address (`<contract:0x945304eb96065b2a98b57a48a06ae28d285a71b5>`) which is also in the `pre` with specified code and a specified `nonce` of `0` so I think I'm either misunderstanding the state of the code or something is still off with resolution and the python-side logic.

```py
>>> pre  # differing nonces here
Alloc(root={
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x10\x00': Account(nonce=1, balance=0, code=b'\x7f\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xfe\x7f\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x7f\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xfe\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\x7f\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc3Pa~v\x08a|\x95\xf3\xffX@v\xe0{ZD`\x00QU', storage=Storage(root={})), 
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\x00': Account(nonce=1, balance=46, code=b'`\x005T\x15`\tW\x00[` 5`\x005U', storage=Storage(root={})), 
b'\xa9OSt\xfc\xe5\xed\xbc\x8e*\x86\x97\xc1S1g~n\xbf\x0b': Account(nonce=0, balance=1000000000000000000, code=b'', storage=Storage(root={}))
})
>>> post  # this looks correct
Alloc(root={
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x10\x00': Account(nonce=0, balance=0, code=b'', storage=Storage(root={})), 
b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\x00': Account(nonce=0, balance=0, code=b'', storage=Storage(root={})), 
b'\xa9OSt\xfc\xe5\xed\xbc\x8e*\x86\x97\xc1S1g~n\xbf\x0b': Account(nonce=1, balance=0, code=b'', storage=Storage(root={}))
})
```

This is a `pdb.set_trace()` at line `159` in `state_static.py`, just before returning `state_test(...)`. You can see two of the addresses have a nonce of `1` in the `pre` but the actual test defines `0`. You can see too the same address as the hard-coded one for sender is still there, though the others looked resolved. I will take a closer look but just wanted to note these discrepancies after a first quick pass of the changes.

fselmo commented:
> You can see too the same address as the hard-coded one for sender is still there, though the others looked resolved.

Actually I think that if we run with the shared pre-alloc we get the achieved unique address resolution and this is what matters. This part seems to work as expected.

fselmo commented on src/ethereum_test_specs/static_state/account.py:
```diff
@@ -1,22 +1,141 @@
+            if compiled_code := self.code.compiled(tags):
+                account_properties["code"] = compiled_code
+        if self.nonce is not None:
+            if self.nonce > 0:
```
This seems to be the fix to the nonce issue I was seeing.

```suggestion
```

fselmo commented on src/ethereum_test_specs/static_state/account.py:
```diff
@@ -1,22 +1,141 @@
+                account_properties["code"] = compiled_code
+        if self.nonce is not None:
+            if self.nonce > 0:
+                account_properties["nonce"] = self.nonce
```
```suggestion
            account_properties["nonce"] = self.nonce
```

fselmo commented:
Ok, I fixed the nonce issue locally but tags inside `raw:` code in `tests/static/state_tests/stRandom/randomStatetest5Filler.json`, for example are not parsing... I either don't get the changes or these are still failing for a lot of cases. I'm going to wait for your initial comments before I go any further as this changed from my flow significantly and things seem to be broken to how I was running them. I will take another pass at this soon to try to understand more of what's failing.

marioevz commented:
> Ok, I fixed the nonce issue locally but tags inside `raw:` code in `tests/static/state_tests/stRandom/randomStatetest5Filler.json`, for example are not parsing... I either don't get the changes or these are still failing for a lot of cases. I'm going to wait for your initial comments before I go any further as this changed from my flow significantly and things seem to be broken to how I was running them. I will take another pass at this soon to try to understand more of what's failing.

Now fixed in https://github.com/fselmo/execution-spec-tests/pull/1/commits/cea6735ef349f91dd789e4f66b33c5790dab8ec1, it was completely broken, sorry about that. The problem was that it was replacing like: `0x1234<contract:...>` -> `0x12340xABCD`, so inserting the `0x` in the middle of the raw code.
